packages/modular-scripts/src/utils/getChangedWorkspaces.ts

```diff
@@ -0,0 +1,65 @@
+import path from 'path';
+import pkgUp from 'pkg-up';
+import { getDiffedFiles } from './gitActions';
+import getModularRoot from './getModularRoot';
+import type { WorkspaceContent } from './getAllWorkspaces';
+
+// Gets a list of changed files, then maps them to their workspace and returns a subset of WorkspaceContent
+export async function getChangedWorkspaces(
+  workspaceContent: WorkspaceContent,
+  targetBranch: string,
+): Promise<WorkspaceContent> {
+  const diffedFiles = getDiffedFiles(targetBranch);
+  const modularRoot = getModularRoot();
+
+  // Resolve each of the changed files to their nearest package.json. The resulting list can contain duplicates and null holes
+  const packageManifestPaths = await Promise.all(
+    diffedFiles.map((changedFile) => pkgUp({ cwd: path.dirname(changedFile) })),
+  );
+
+  return matchWorkspaces(packageManifestPaths, modularRoot, workspaceContent);
+}
+
+// Match workspace manifest paths to a subset of WorkspaceContent. This function works completely in memory and is test-friendly
+export function matchWorkspaces(
+  packagePaths: (string | null)[],
+  root: string,
+  workspaceContent: WorkspaceContent,
+): WorkspaceContent {
+  const [packages, workspaces] = workspaceContent;
+  const workspaceEntries = Object.entries(workspaces);
+  const result: WorkspaceContent = [new Map([]), {}];
+
+  for (const packagePath of packagePaths) {
+    const [resultPackages, resultWorkspaces] = result;
+    // Ignore holes
+    if (!packagePath) continue;
+    // Get the package directory from the package.json path and make it relative to the root, for comparison with the original WorkspaceContent
+    const packageDir = path.relative(root, path.dirname(packagePath));
```

Comment:
Should this check that the `packageDir` is a child of the `root` to prevent the possibility of finding `package.json` files outside of the repo

Reply:
I don't think it's possible; all the changed files that we find are inside the repo by definition (as we use the `git` command to find them), so we will _at least_ find the root `package.json`. If there's no root manifest, then the repo is not a modular repo and is not a monorepo, so we will never get here.

Reply:
👍